includes/utils/class-amp-mutation-utils.php

```diff
@@ -0,0 +1,215 @@
+	 *
+	 * @param array  $histogram The count of attributes or nodes removed.
+	 * @param string $key The attribute or node name removed.
+	 * @return array $histogram The incremented histogram.
+	 */
+	public static function increment_count( $histogram, $key ) {
+		$current_value     = isset( $histogram[ $key ] ) ? $histogram[ $key ] : 0;
+		$histogram[ $key ] = $current_value + 1;
+		return $histogram;
+	}
+
+	/**
+	 * Gets whether a node was removed in a sanitizer.
+	 *
+	 * @return boolean.
+	 */
+	public static function was_node_removed() {
+		return ! empty( self::$removed_nodes );
+	}
+
+	/**
+	 * Processes markup, to determine AMP validity.
+	 *
+	 * Passes $markup through the AMP sanitizers.
+	 * Also passes a 'mutation_callback' to keep track of stripped attributes and nodes.
+	 *
+	 * @param string $markup The markup to process.
+	 * @return void.
+	 */
+	public static function process_markup( $markup ) {
+		$args = array(
+			'content_max_width' => ! empty( $content_width ) ? $content_width : AMP_Post_Template::CONTENT_MAX_WIDTH,
+			'mutation_callback' => 'AMP_Mutation_Utils::track_removed',
+		);
+		AMP_Content_Sanitizer::sanitize( $markup, amp_get_content_sanitizers(), $args );
+	}
+
+	/**
+	 * Registers the REST API endpoint for validation.
+	 *
+	 * @return void.
+	 */
+	public static function amp_rest_validation() {
+		register_rest_route( 'amp-wp/v1', '/validate', array(
+			'methods'             => 'POST',
+			'callback'            => array( __CLASS__, 'validate_markup' ),
+			'args'                => array(
+				self::MARKUP_KEY => array(
+					'validate_callback' => array( __CLASS__, 'validate_arg' ),
+				),
+			),
+			'permission_callback' => array( __CLASS__, 'permission' ),
+		) );
+	}
+
+	/**
+	 * The permission callback for the REST request.
+	 *
+	 * @return boolean $has_permission Whether the current user has the permission.
+	 */
+	public static function permission() {
+		return current_user_can( 'edit_posts' );
+	}
+
+	/**
+	 * Validate the markup passed to the REST API.
+	 *
+	 * @param WP_REST_Request $request The REST request.
+	 * @return array|WP_Error.
+	 */
+	public static function validate_markup( WP_REST_Request $request ) {
+		$json = $request->get_json_params();
+		if ( empty( $json[ self::MARKUP_KEY ] ) ) {
+			return new WP_Error( 'no_markup', 'No markup passed to validator', array(
+				'status' => 404,
+			) );
+		}
+
+		return self::get_response( $json[ self::MARKUP_KEY ] );
+	}
+
+	/**
+	 * Gets the AMP validation response.
+	 *
+	 * If $markup isn't passed,
+	 * It will return the validation errors the sanitizers found in rendering the page.
+	 *
+	 * @param string $markup To validate for AMP compatibility (optional).
+	 * @return array $response The AMP validity of the markup.
+	 */
+	public static function get_response( $markup = null ) {
+		$response = array();
+		if ( isset( $markup ) ) {
+			self::process_markup( $markup );
+			$response['processed_markup'] = esc_html( $markup );
```

Comment:
`esc_html()` probably isn't the best solution, but something is needed. Without it, almost the whole tag is stripped. Here's the result of this from Postman:
```
{"has_error":false,"removed_nodes":null,"removed_attributes":{"onchange":1},"processed_markup":"<amp-youtube onchange='someFunction()' data-videoid='XOY3ZUO6P0k' layout='responsive' width='694' height='390'><\/amp-youtube>"}
```

It looks like it escapes `/` characters:
```
<\/amp-youtube>
```